jlib/timeparse/timeparse.go

```diff
@@ -0,0 +1,162 @@
+package timeparse
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+)
+
+// DateDim is the date dimension object returned from the timeparse function
+type DateDim struct {
+	DateID         string `json:"DateId"`
+	DateKey        string `json:"DateKey"`
+	UTC            string `json:"UTC"`
+	DateUTC        string `json:"DateUTC"`
+	Parsed         string `json:"Parsed"`
+	Local          string `json:"Local"`
+	DateLocal      string `json:"DateLocal"`
+	HourID         string `json:"HourId"`
+	HourKey        string `json:"HourKey"`
+	Millis         string `json:"Millis"`
+	Hour           string `json:"Hour"`
+	TimeZone       string `json:"TimeZone"`
+	TimeZoneOffset string `json:"TimeZoneOffset"`
+	YearMonth      string `json:"YearMonth"`
+	YearWeek       string `json:"YearWeek"`
+	YearIsoWeek    string `json:"YearIsoWeek"`
+	YearDay        string `json:"YearDay"`
+
+	// TODO add UTC fields
+}
+
+// TimeDateDimensions generates a JSON object dependent on input source timestamp, input source format and input source timezone
+// using golang time formats
+func TimeDateDimensions(inputSrcTs, inputSrcFormat, inputSrcTz string) (interface{}, error) {
+	location, err := time.LoadLocation(inputSrcTz)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO clean up, speed up, make more efficient etc etc - but first, get it to work!
+
+	parsedTime, err := parseDateTimeLocation(inputSrcTs, inputSrcFormat, location)
+	if err != nil {
+		return nil, err
+	}
+
+	// convert the parsed time into a UTC time for UTC calculations
+	parsedTimeUTC := parsedTime.UTC()
+
+	dateDim := DateDim{}
+
+	// UTC TIME values
+
+	dateDim.UTC = parsedTimeUTC.Format("2006-01-02T15:04:05.000Z")
```

Comment:
Might be a bit clearer if you created `dateDim` via some form of constructor or maybe assigned each value to the struct in one go

```
DateDim {
    DateUTC : parsedTimeUTC.Format("2006-01-02"),
    DateID : "Dates_" + dateID
}
```

etc